### PR TITLE
planner: plan cache support subquery

### PR DESCRIPTION
**Pull request by fzzf678**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)

Starting from v7.0.0, prepared plan cache supports subquery.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.4 (TiDB 6.4 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):
https://github.com/pingcap/tidb/issues/40219
### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: sql-prepared-plan-cache.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ TiDB 优化器对这两类查询的处理是一样的：`Prepare` 时将参数
 
 - `SELECT`、`UPDATE`、`INSERT`、`DELETE`、`Union`、`Intersect`、`Except` 以外的 SQL 语句；
 - 访问分区表、临时表或访问表中包含生成列的查询；
-- 包含子查询的查询，如 `select * from t where a > (select ...)`；
+- 查询中包含非关联子查询或无法解关联的关联子查询；
```
Comment (qw4990) [inline sql-prepared-plan-cache.md:21]: 需要举一个例子来说明下 关联子查询 和 非关联子查询

Comment (fzzf678) [reply sql-prepared-plan-cache.md:21]:
```suggestion
- 查询中包含非关联子查询或无法解关联的关联子查询，例如 `select * from t1 where t1.a > (select 1 from t2 where t2.b < 1)` 和 `select * from t1 where t1.a > (select a from t2 where t1.b > t2.b)` 分别是无法缓存的非关联子查询和关联子查询
```

File: sql-prepared-plan-cache.md [hunk 2]
```diff
@@ -18,7 +18,7 @@ TiDB 优化器对这两类查询的处理是一样的：`Prepare` 时将参数
 
 - `SELECT`、`UPDATE`、`INSERT`、`DELETE`、`Union`、`Intersect`、`Except` 以外的 SQL 语句；
 - 访问分区表、临时表或访问表中包含生成列的查询；
-- 包含子查询的查询，如 `select * from t where a > (select ...)`；
+- 查询中包含非关联子查询或无法解关联的关联子查询，例如 `select * from t1 where t1.a > (select 1 from t2 where t2.b < 1)` 和 `select * from t1 where t1.a > (select a from t2 where t1.b > t2.b)` 分别是无法缓存的非关联子查询和关联子查询
```
Comment (fzzf678) [inline sql-prepared-plan-cache.md:21]:
```suggestion
- 查询中包含非关联子查询的，例如 `select * from t1 where t1.a > (select 1 from t2 where t2.b < 1)` 
- 执行计划中带有 `PhysicalApply` 算子的关联子查询，例如  `select * from t1 where t1.a > (select a from t2 where t1.b > t2.b)` 
```

File: sql-prepared-plan-cache.md [hunk 3]
```diff
@@ -18,7 +18,8 @@ TiDB 优化器对这两类查询的处理是一样的：`Prepare` 时将参数
 
 - `SELECT`、`UPDATE`、`INSERT`、`DELETE`、`Union`、`Intersect`、`Except` 以外的 SQL 语句；
 - 访问分区表、临时表或访问表中包含生成列的查询；
-- 包含子查询的查询，如 `select * from t where a > (select ...)`；
+- 查询中包含非关联子查询的，例如 `select * from t1 where t1.a > (select 1 from t2 where t2.b < 1)`
```
Comment (fzzf678) [inline sql-prepared-plan-cache.md:21]:
```suggestion
- 查询中包含非关联子查询的，例如 `select * from t1 where t1.a > (select 1 from t2 where t2.b < 1)` ；
```

File: sql-prepared-plan-cache.md [hunk 4]
```diff
@@ -18,7 +18,8 @@ TiDB 优化器对这两类查询的处理是一样的：`Prepare` 时将参数
 
 - `SELECT`、`UPDATE`、`INSERT`、`DELETE`、`Union`、`Intersect`、`Except` 以外的 SQL 语句；
 - 访问分区表、临时表或访问表中包含生成列的查询；
-- 包含子查询的查询，如 `select * from t where a > (select ...)`；
+- 查询中包含非关联子查询的，例如 `select * from t1 where t1.a > (select 1 from t2 where t2.b < 1)` 
+- 执行计划中带有 `PhysicalApply` 算子的关联子查询，例如  `select * from t1 where t1.a > (select a from t2 where t1.b > t2.b)`
```
Comment (fzzf678) [inline sql-prepared-plan-cache.md:22]:
```suggestion
- 执行计划中带有 `PhysicalApply` 算子的关联子查询，例如  `select * from t1 where t1.a > (select a from t2 where t1.b > t2.b)` ；
```

File: sql-prepared-plan-cache.md [hunk 5]
```diff
@@ -18,7 +18,8 @@ TiDB 优化器对这两类查询的处理是一样的：`Prepare` 时将参数
 
 - `SELECT`、`UPDATE`、`INSERT`、`DELETE`、`Union`、`Intersect`、`Except` 以外的 SQL 语句；
 - 访问分区表、临时表或访问表中包含生成列的查询；
-- 包含子查询的查询，如 `select * from t where a > (select ...)`；
+- 查询中包含非关联子查询的，例如 `select * from t1 where t1.a > (select 1 from t2 where t2.b < 1)` ；
+- 执行计划中带有 `PhysicalApply` 算子的关联子查询，例如  `select * from t1 where t1.a > (select a from t2 where t1.b > t2.b)` ；
```
Comment (fzzf678) [inline sql-prepared-plan-cache.md:22]:
```suggestion
- 查询中包含非关联子查询的，例如 `select * from t1 where t1.a > (select 1 from t2 where t2.b < 1)`；
- 执行计划中带有 `PhysicalApply` 算子的关联子查询，例如  `select * from t1 where t1.a > (select a from t2 where t1.b > t2.b)`；
```

**Review by qw4990 — APPROVED** [review]
LGTM

**Review by chrysan — APPROVED** [review]

File: sql-prepared-plan-cache.md [hunk 6]
```diff
@@ -18,7 +18,8 @@ TiDB 优化器对这两类查询的处理是一样的：`Prepare` 时将参数
 
 - `SELECT`、`UPDATE`、`INSERT`、`DELETE`、`Union`、`Intersect`、`Except` 以外的 SQL 语句；
 - 访问分区表、临时表或访问表中包含生成列的查询；
-- 包含子查询的查询，如 `select * from t where a > (select ...)`；
+- 查询中包含非关联子查询的，例如 `select * from t1 where t1.a > (select 1 from t2 where t2.b < 1)`；
+- 执行计划中带有 `PhysicalApply` 算子的关联子查询，例如  `select * from t1 where t1.a > (select a from t2 where t1.b > t2.b)`；
```
Comment (Oreoxmt) [inline sql-prepared-plan-cache.md:22]:
```suggestion
- 查询中包含非关联子查询，例如 `SELECT * FROM t1 WHERE t1.a > (SELECT 1 FROM t2 WHERE t2.b < 1)`；
- 执行计划中带有 `PhysicalApply` 算子的关联子查询，例如 `SELECT * FROM t1 WHERE t1.a > (SELECT a FROM t2 WHERE t1.b > t2.b)`；
```

**Review by Oreoxmt — APPROVED** [review]
Rest LGTM

**Review by ran-huang — APPROVED** [review]

File: system-variables.md [hunk 7]
```diff
@@ -1533,6 +1533,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
 - 默认值：`ON`
 - 这个变量用来控制 Prepared Plan Cache 是否缓存 `LIMIT` 后面带变量 (`LIMIT ?`) 的执行计划。目前不支持缓存 `LIMIT` 后面带变量且变量值大于 10000 的执行计划。
 
+### `tidb_enable_plan_cache_for_subquery` <span class="version-mark">从 v7.0.0 版本开始引入</span>
```
Comment (Oreoxmt) [inline system-variables.md:1536]: @ran-huang PTAL

**Review by qw4990 — APPROVED** [review]

**Review by ran-huang — APPROVED** [review]

**Pull request merged**